src/commissioner/ExampleCommissioningStateMachine.cpp

```diff
@@ -0,0 +1,306 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <commissioner/ExampleCommissioningStateMachine.h>
+
+namespace chip {
+namespace Commissioner {
+namespace ExampleCommissioningStateMachine {
+
+void StateFactory::Init(OpCredsIssuer * issuer, FabricIndex fabricIndex, NodeId nodeId, ByteSpan operationalDataset, ByteSpan ssid,
+                        ByteSpan wiFiCredentials)
+{
+    mIssuer             = issuer;
+    mFabricIndex        = fabricIndex;
+    mNodeId             = nodeId;
+    mOperationalDataset = operationalDataset;
+    mSsid               = ssid;
+    mWiFiCredentials    = wiFiCredentials;
+}
+
+void StateFactory::SetCallbacks(OnSuccess onSuccess, OnFailure onFailure)
+{
+    mOnSuccess = onSuccess;
+    mOnFailure = onFailure;
+}
+
+chip::StateMachine::Optional<State> Transitions::operator()(const State & state, const Event & event)
+{
+    if (state.Is<Idle>() && event.Is<OnboardingPayload>())
+    {
+        return mFactory.CreateParsingOnboardingPayload(event.Get<OnboardingPayload>());
+    }
+    else if ((state.Is<Idle>() || state.Is<ParsingOnboardingPayload>()) && event.Is<ParsedPayload>())
+    {
+        CHIP_ERROR err = mCtx.ScheduleTimeout(kCommissionableDiscoveryTimeout);
```

Comment:
Could `kCommissionableDiscoveryTimeout` be better served as an argument into this state machine if certain platforms wanted to have different timeouts?

Reply:
Actually, this is managed completely within `ExampleCommissioningStateMachine` expressly to support platform-specific behavior.

Notice that none of the core state logic in `src/commissioner/States.h` or `src/commissioner/Events.h` has any notion of the timeout.  This is a very purposeful component of the design where, yes, platforms will definitely want to control this and can do so with their own state machine in place of `ExampleCommissioningStateMachine`.

This brings up an important aspect of the design.  `ExampleCommissioningStateMachine` is not this feature's interface.  This feature's interface is `src/commissioner/States.h`.  `ExampleCommissioningStateMachine` is the shim that `chip-tool` is specifically using to interact with `src/commissioner/States.h`, and is here to serve as a fully-worked example for how other platforms can interact with `src/commissioner/States.h`.